### PR TITLE
/EBCS/INLET : new inlet condition (ALE staggered upwind scheme)

### PR DESCRIPTION
**Pull request by Schwoertzig**
#### Description of the changes
Boundary Condition INLET

Verification done with test case " Mach3 windtunnel with a step " 
- Law6
- Law51 single phase
- Law51 multi phases
- 2D / 3D
- VE (Vel,rho,e) and VP (Vel,rho,P) formulations
- Vn input instead Vx,Vy,Vz
-  SMPD / SPMD
- RESTART (*1.rad -> *2.rad)

VP formulation : Impose Velocity and Pressure (and density)
VE formulation : Impose Velocity and Energy (and density)

Verification Test :
image

image
image

.
Input File (Verification Test)
[test-case--MACH3.zip](https://github.com/user-attachments/files/22865949/test-case--MACH3.zip)

Input Format (example with 3 submaterials) : 
image

### CODE REVIEW

File: engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90 [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+          WNF = ZERO
+          IF(IALE == 1)THEN
+            WMX=W(1,NG1)+w(1,NG2)+w(1,NG3)
+            WMY=W(2,NG1)+w(2,NG2)+w(2,NG3)
+            WMZ=W(3,NG1)+w(3,NG2)+w(3,NG3)
+            FAC = THIRD
+            IF(N4 /= N3) THEN
+               WMX = WMX+V(1,NG4)
+               WMY = WMY+V(2,NG4)
+               WMZ = WMZ+V(3,NG4)
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90]: good catch. fixed

File: engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90 [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+          LA(3,N1)=LA(3,N1)+NZ
+          LA(2,N2)=LA(2,N2)+NY
+          LA(3,N2)=LA(3,N2)+NZ
+          YM = ONE
+          IF(N2D==1)YM=HALF*(X(2,NG1)+X(2,NG2))
+
+          !CUMULATIVE AREA FOR MEAN IMPEDENCE (NUMERATOR)
+          NORM = SQRT(NY*NY + NZ*NZ)
+          AREA = NORM*YM
+          EBCS%AREA = AREA
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90]: good catch. This passed the verification test because all faces had same area. fixed !

File: starter/source/boundary_conditions/ebcs/iniebcs.F [hunk 3]
```diff
@@ -265,6 +265,28 @@ SUBROUTINE INIEBCS(ALE_CONNECTIVITY, IFLAG,IGRSURF, IXS, IXQ, IXTG,
...
+               EBCS%dvnf(1:EBCS%nb_elem) = ZERO
+               !     ng
+               EBCS%has_ng = .true.
+               IF(IFLAG==0)ALLOCATE(EBCS%ng(EBCS%nb_elem))
+               EBCS%ng(1:EBCS%nb_elem) = 0
+               !     iloc
+               EBCS%has_iloc = .true.
+               IF(IFLAG==0)ALLOCATE(EBCS%iloc(EBCS%nb_elem))
+               EBCS%iloc(1:EBCS%nb_elem) = 0
+             ENDIF
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply starter/source/boundary_conditions/ebcs/iniebcs.F:289]: False Warning. Nothing to do.
Staggered scheme has an independent treatment.

File: engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90 [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+      real(kind=WP), intent(in) :: BUFMAT(*)
+      INTEGER,INTENT(IN) :: NCYCLE
+! ----------------------------------------------------------------------------------------------------------------------
+!                                                   Local Variables
+! ----------------------------------------------------------------------------------------------------------------------
+      INTEGER :: I,IS,KSEG,N1,N2,N3,N4,NG1,NG2,NG3,NG4,N
+      REAL(KIND=WP) :: ORIENT,FAC,X13,Y13,Z13,X24,Y24,Z24,NX,NY,NZ,S !< tmp variables (normal + mean velocities)
+      REAL(KIND=WP) :: DVNF,VNF,WNF !< normal veloicty at face : relative(dv), material(v) & grid(w).
+      REAL(KIND=WP) :: WMX,WMY,WMZ !< mean grid velocity at face
+      REAL(KIND=WP) :: FLUXI,FLUXO !< voluemtrix fluxes (in/out)
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90]: typo fixed

File: engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90 [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+        M_out = M_out-FLUXO*RHOII
+        E_in  = E_in-FLUXI*EINTJJ
+        E_out = E_out-FLUXO*EINTII
+        
+        !/TH/SURF (massflow, velocity, pressure)
+        PFACE=PJJ
+        IF(DVNF > ZERO)PFACE=PII
+        FSAVSURF(2,SURF_ID) = FSAVSURF(2,SURF_ID) + RHOII*AREA*DVNF     ! rho.S.u = dm/dt
+        FSAVSURF(3,SURF_ID) = FSAVSURF(3,SURF_ID) + AREA*DVNF           ! S.u
+        FSAVSURF(4,SURF_ID) = FSAVSURF(4,SURF_ID) + AREA*PII            ! S.P
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply engine/source/boundary_conditions/ebcs/ebcs8_inlet.F90]: fixed

File: starter/source/boundary_conditions/ebcs/hm_read_ebcs_inlet.F [hunk 6]
```diff
@@ -360,22 +363,19 @@ SUBROUTINE HM_READ_EBCS_INLET(IGRSURF,NPC, MULTI_FVM, UNITAB, ID, TITR, UID, LSU
          EBCS%fvm_inlet_data%VAL_ALPHA(IMAT)  = ALPHA
          EBCS%fvm_inlet_data%VAL_RHO(IMAT)    = RHO
          EBCS%fvm_inlet_data%VAL_PRES(IMAT)   = PRES
+         EBCS%NBMAT = NBMAT
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply starter/source/boundary_conditions/ebcs/hm_read_ebcs_inlet.F]: moved outside the loop

File: common_source/modules/boundary_conditions/ebcs_mod.F90 [hunk 7]
```diff
@@ -822,8 +860,15 @@ subroutine read_common_data(this)
           if(integer_data(7) == 1)this%has_ielem=.true.
           this%has_th=.false.
           if(integer_data(8) == 1)this%has_th=.true.
+          this%has_ng=.false.
+          if(integer_data(9) == 1)this%has_ng=.true.
+          if(this%has_ng)allocate(this%ng(this%nb_elem))
+          this%has_iloc=.false.
+          if(integer_data(10) == 1)this%has_iloc=.true.
+          if(this%has_iloc)allocate(this%iloc(this%nb_elem))
```
[the remark this answers is not in the document]
Comment (Schwoertzig) [reply common_source/modules/boundary_conditions/ebcs_mod.F90:868]: Good remark but False positive. They are not written in restart because it can be computed at the beggining of the run. This way it does not increase the size of restart file.

**Comment by sebastienVilleneuve** [comment]
@Schwoertzig 

You have 2 commits in this PR , could you squash them ?

Thanks,
Sébastien

**Comment by Schwoertzig** [comment]
> @Schwoertzig
> 
> You have 2 commits in this PR , could you squash them ?
> 
> Thanks, Sébastien

Fi @sebastienVilleneuve 
It is Done.
Regards
